File: Leetcode-solution-c-c++/longestCycleInAGraph.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
    int dfs(int node, vector<int> &edges, vector<bool> &vis, int depth, int par)
    {
        if (node == -1)
        {
            return -1;
        }
        if (vis[node])
        {
            return (node == par) ? depth : -1;
        }
        vis[node] = true;
        int temp = dfs(edges[node], edges, vis, depth+1, par);
        if(temp == -1)
        {
            vis[node] = false;
        }
        return temp;
    }

public:
    int longestCycle(vector<int> &edges)
    {
        int n = edges.size();
        int ans = -1;
        vector<bool> vis(n, false);

        for (int node = 0; node < n; node++)
        {
            if (vis[node] || edges[node] == -1)
            {
                continue;
            }
            ans = max(ans, dfs(node, edges, vis, 0, node));
        }
        return ans;
    }
};

File: Leetcode-solution-python/longestCycleInAGraph.py
from typing import List


class Solution:

    def dfs(self, node: int, edges: List[int], vis: List[bool], depth: int, par: int) -> int:
        if node == -1:
            return -1
        if vis[node]:
            if node == par:
                return depth
            else:
                return -1
        vis[node] = True
        temp = self.dfs(edges[node], edges, vis, depth + 1, par)
        if temp == -1:
            vis[node] = False
        return temp

    def longestCycle(self, edges: List[int]) -> int:
        ans, n = -1, len(edges)
        vis = [False] * n
        for i in range(n):
            if vis[i] or edges[i] == -1:
                continue
            ans = max(ans, self.dfs(i, edges, vis, 0, i))
        return ans
